### PR TITLE
[T04]: Search job by title

### PR DESCRIPTION
**Pull request by ChienVan01**
Stories:
Link code feature: https://github.com/fabatek/onboarding-job-board-frontend/tree/feature/chientv/t04_search_job_by_title
Link deploy surge: https://delirious-seat.surge.sh/
Link API to see specific job totals: https://6239311d043817a543df0a3f.mockapi.io/api/jobs

Notes: 

1. Add SEARCH_FILTER type in JobsActionType.ts
2. Add case SEARCH_FILTER in JobReducer.ts
3. Edit Function GetJob() in JobsAction.ts
4. Add Loading button in JobList component
5. Add unit test in Home page

results
image
image

### CODE REVIEW

File: src/actions/JobsAction.ts [hunk 1]
```diff
@@ -4,25 +4,39 @@ import {
...
-    );
+export const GetJob =
+  (searchTitle: string) =>
+  async (dispatch: Dispatch<JobDispatchTypes>): Promise<void> => {
+    try {
+      dispatch({
+        type: JOBS_LOADING,
+      });
+      const res = await axios.get(
+        `https://6239311d043817a543df0a3f.mockapi.io/api/jobs`
```
Comment (ducdmfaba) [inline src/actions/JobsAction.ts:19]: Please extract `https://6239311d043817a543df0a3f.mockapi.io/api` to constant.

File: src/actions/JobsAction.ts [hunk 2]
```diff
@@ -4,25 +4,39 @@ import {
...
+        : dispatch({
+            type: SEARCH_FILTER,
+            payload: {
+              search: searchTitle,
+              jobs: res.data.filter(
+                (i: { name: string | string[] }) =>
+                  i.name.indexOf(searchTitle) >= 0
+              ),
+            },
+          });
```
Comment (ducdmfaba) [inline src/actions/JobsAction.ts:36]: Why did you use filter? I think the endpoint currently supported search functionality.

File: src/components/jobList/JobList.tsx [hunk 3]
```diff
@@ -28,7 +28,18 @@ const JobList: React.FC = () => {
...
+    <div className='py-5 d-flex justify-content-center'>
+      <button className='btn btn-primary ' type='button' disabled>
+        <span
+          className='spinner-border spinner-border-sm'
+          role='status'
+          aria-hidden='true'
+        ></span>
+        Loading...
+      </button>
+    </div>
```
Comment (ducdmfaba) [inline src/components/jobList/JobList.tsx:41]: We should have a component `Loading`

File: src/components/seachContainer/SearchContainer.tsx [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
+import { useDispatch, useSelector } from 'react-redux';
 import { RootStore } from '../../store';
-import React, { ReactElement } from 'react';
+import React, { ReactElement, useState } from 'react';
+import { GetJob } from '../../actions/JobsAction';
 
 const SearchContainer: React.FC = () => {
   const jobState = useSelector((state: RootStore) => state.job);
+  const dispatch = useDispatch();
+  const [searchText, setSearchText] = useState('');
```
Comment (ducdmfaba) [inline src/components/seachContainer/SearchContainer.tsx:11]: Missing data type.

File: src/containers/Home/index.tsx [hunk 5]
```diff
@@ -16,7 +16,7 @@ const Home: React.FC = () => {
   const jobState = useSelector((state: RootStore) => state.job);
 
   useEffect(() => {
-    dispatch(GetJob());
+    dispatch(GetJob(''));
```
Comment (ducdmfaba) [inline src/containers/Home/index.tsx:19]: The function name should be `getJobs`

File: src/containers/Home/index.tsx [hunk 6]
```diff
@@ -16,7 +16,7 @@ const Home: React.FC = () => {
...
   }, []);
 
   console.log(jobState);
```
Comment (ducdmfaba) [inline src/containers/Home/index.tsx:22]: Remove test code.

**Review by ducdmfaba — CHANGES_REQUESTED** [review]

**Comment by ducdmfaba** [comment]
@ChienVan01 Please let me know when you have finished this.

**Comment by ChienVan01** [comment]
> @ChienVan01 Please let me know when you have finished this.

Sorry, I'll finish it today

File: src/actions/JobsAction.ts [hunk 7]
```diff
@@ -4,25 +4,36 @@ import {
...
+export const getJob =
+  (searchTitle: string) =>
+  async (dispatch: Dispatch<JobDispatchTypes>): Promise<void> => {
+    try {
+      dispatch({
+        type: JOBS_LOADING,
+      });
+      const res = await axios.get(BASE_URL + `?name=${searchTitle}`);
+
+      searchTitle === ''
```
Comment (ducdmfaba) [inline src/actions/JobsAction.ts:22]: Just `!!searchTitle`

File: src/actions/JobsAction.ts [hunk 8]
```diff
@@ -4,25 +4,36 @@ import {
...
-  }
-};
+export const getJob =
+  (searchTitle: string) =>
+  async (dispatch: Dispatch<JobDispatchTypes>): Promise<void> => {
+    try {
+      dispatch({
+        type: JOBS_LOADING,
+      });
+      const res = await axios.get(BASE_URL + `?name=${searchTitle}`);
```
Comment (ducdmfaba) [inline src/actions/JobsAction.ts:20]: - Please use `template string` for this. 
- When `searchTitle` is null, endpoint should be `BASE_URL` instead `BASE_URL?name=`

File: src/actions/JobsAction.ts [hunk 9]
```diff
@@ -4,25 +4,36 @@ import {
...
+
+      searchTitle === ''
+        ? dispatch({
+            type: JOBS_SUCCESS,
+            payload: res.data,
+          })
+        : dispatch({
+            type: SEARCH_FILTER,
+            payload: {
+              search: searchTitle,
```
Comment (ducdmfaba) [inline src/actions/JobsAction.ts:30]: Do we need to store `searchTitle`?

File: src/components/jobList/JobList.tsx [hunk 10]
```diff
@@ -12,25 +12,38 @@ const JOB_PER_PAGE = 10;
...
+        <button className='btn btn-primary' type='button' disabled>
+          <span
+            className='spinner-border spinner-border-sm'
+            role='status'
+            aria-hidden='true'
+          ></span>
+          Loading...
+        </button>
+      </div>
+    </>
```
Comment (ducdmfaba) [inline src/components/jobList/JobList.tsx:43]: Please extract this to the component. If we have existed, try to re-use it.

File: src/components/searchContainer/SearchContainer.tsx [hunk 11]
```diff
@@ -35,6 +38,13 @@ const SearchContainer: React.FC = () => {
     );
   };
 
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    console.log(e.target.value);
```
Comment (ducdmfaba) [inline src/components/searchContainer/SearchContainer.tsx:42]: Remove test code.

File: src/containers/Home/index.test.tsx [hunk 12]
```diff
@@ -1,11 +1,39 @@
...
+
+describe('Should test get data API', () => {
+  it('Get all data API request successfully', async () => {
+    const result = await axios.get(BASE_URL);
+    expect(result.data.length).toBe(100);
+  });
+
+  it('get data API with endpoint request successfully', async () => {
+    const itemSearch: string = 'name 100';
+    const result = await axios.get(BASE_URL + `?name=${itemSearch}`);
```
Comment (ducdmfaba) [inline src/containers/Home/index.test.tsx:36]: `template string`

File: src/containers/Home/index.tsx [hunk 13]
```diff
@@ -1,25 +1,18 @@
...
 const Home: React.FC = () => {
   const dispatch = useDispatch();
-  const jobState = useSelector((state: RootStore) => state.job);
 
   useEffect(() => {
-    dispatch(GetJob());
-  }, []);
-
-  console.log(jobState);
+    dispatch(getJob(''));
```
Comment (ducdmfaba) [inline src/containers/Home/index.tsx:14]: No need empty string for argument here.

**Review by ducdmfaba — CHANGES_REQUESTED** [review]

File: src/components/jobList/JobList.tsx [hunk 14]
```diff
@@ -12,25 +12,42 @@ const JOB_PER_PAGE = 10;
...
```
Comment (ducdmfaba) [inline src/components/jobList/JobList.tsx:43]: Please move it outside the component to reuse.

**Review by ducdmfaba — CHANGES_REQUESTED** [review]

File: src/actions/JobsAction.ts (hunk 8 above)
Comment (ducdmfaba) [reply src/actions/JobsAction.ts]: Is it fixed?

File: src/components/jobList/JobList.tsx [hunk 15]
```diff
@@ -71,4 +73,19 @@ const JobList: React.FC = () => {
...
+          className='spinner-border spinner-border-sm'
+          role='status'
+          aria-hidden='true'
+        ></span>
+        Loading...
+      </button>
+    </div>
+  </>
+);
+
```
Comment (ducdmfaba) [inline src/components/jobList/JobList.tsx:90]: I mean you should create it in components folders.

**Review by ducdmfaba — CHANGES_REQUESTED** [review]

File: src/components/searchContainer/SearchContainer.tsx [hunk 16]
```diff
@@ -35,6 +38,12 @@ const SearchContainer: React.FC = () => {
     );
   };
 
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    return setSearchText(e.target.value);
```
Comment (ducdmfaba) [inline src/components/searchContainer/SearchContainer.tsx:42]: Why did you use function return for this?

File: src/components/jobList/JobList.tsx [hunk 17]
```diff
@@ -1,36 +1,43 @@
...
+  useEffect(() => {
+    setCurrentPageNumber(0);
+  }, [jobState.job]);
+
+  return jobState.loading ? (
+    <Loading />
+  ) : jobState.job == null ? (
+    <div className='container w-50 alert alert-danger' role='alert'>
+      ERROR
+    </div>
```
Comment (ducdmfaba) [inline src/components/jobList/JobList.tsx:37]: This does not make sense. 
- I don't think `job` can be null.
- In javascript or typescript avoid equaling by `==`. 
Please remove it.

**Review by ducdmfaba — CHANGES_REQUESTED** [review]

**Review by ducdmfaba — APPROVED** [review]

**Pull request merged**